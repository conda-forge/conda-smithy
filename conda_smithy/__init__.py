try:
    from conda_smithy._version import __version__
except ImportError:
    __version__ = "0.0.0"
