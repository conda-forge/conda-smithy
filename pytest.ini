[pytest]
markers =
    legacy_circle: Test designed to run as if prior to the azure migration
    legacy_travis: Test designed to run as if prior to the azure migration
    legacy_appveyor: Test designed to run as if prior to the azure migration
    cli: CLI tests outside of test/test_cli.py