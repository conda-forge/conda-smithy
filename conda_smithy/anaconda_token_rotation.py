"""This module updates/rotates anaconda/binstar tokens.

The correct way to use this module is to call its functions via the command
line utility. The relevant one is

    conda-smithy update-anaconda-token

Note that if you are using appveyor, you will need to push the changes to the
conda-forge.yml in your feedstock to GitHub.
"""

import os
import sys
from contextlib import redirect_stderr, redirect_stdout
from typing import Optional, Union

import requests
from github import Github

from conda_smithy.utils import update_conda_forge_config


def _get_anaconda_token():
    """use this helper to enable easier patching for tests"""
    try:
        from conda_smithy.ci_register import anaconda_token

        return anaconda_token
    except ImportError:
        raise RuntimeError(
            "You must have the anaconda token defined to do token rotation!"
        )


def rotate_anaconda_token(
    user: str,
    project: str,
    feedstock_config_path: Optional[str],
    drone: bool = True,
    circle: bool = True,
    travis: bool = True,
    azure: bool = True,
    appveyor: bool = True,
    github_actions: bool = True,
    token_name: str = "BINSTAR_TOKEN",
    drone_endpoints: Union[list, tuple] = (),
):
    """Rotate the anaconda (binstar) token used by the CI providers

    All exceptions are swallowed and stdout/stderr from this function is
    redirected to `/dev/null`. Sanitized error messages are
    displayed at the end.

    If you need to debug this function, define `DEBUG_ANACONDA_TOKENS` in
    your environment before calling this function.
    """
    # we are swallong all of the logs below, so we do a test import here
    # to generate the proper errors for missing tokens
    # note that these imports cover all providers
    from .ci_register import travis_endpoint  # noqa
    from .azure_ci_utils import default_config  # noqa
    from conda_smithy.github import gh_token

    anaconda_token = _get_anaconda_token()

    if github_actions:
        gh = Github(gh_token())

    # capture stdout, stderr and suppress all exceptions so we don't
    # spill tokens
    failed = False
    err_msg = None
    with open(os.devnull, "w") as fp:
        if "DEBUG_ANACONDA_TOKENS" in os.environ:
            fpo = sys.stdout
            fpe = sys.stderr
        else:
            fpo = fp
            fpe = fp

        with redirect_stdout(fpo), redirect_stderr(fpe):
            try:
                if circle:
                    try:
                        rotate_token_in_circle(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on circle!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if drone:
                    for drone_endpoint in drone_endpoints:
                        try:
                            rotate_token_in_drone(
                                user,
                                project,
                                anaconda_token,
                                token_name,
                                drone_endpoint,
                            )
                        except Exception as e:
                            if "DEBUG_ANACONDA_TOKENS" in os.environ:
                                raise e
                            else:
                                err_msg = (
                                    f"Failed to rotate token for {user}/{project}"
                                    f" on drone endpoint {drone_endpoint}!"
                                )
                                failed = True
                                raise RuntimeError(err_msg)

                if travis:
                    try:
                        rotate_token_in_travis(
                            user,
                            project,
                            feedstock_config_path,
                            anaconda_token,
                            token_name,
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on travis!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if azure:
                    try:
                        rotate_token_in_azure(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on azure!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if appveyor:
                    try:
                        rotate_token_in_appveyor(
                            feedstock_config_path, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on appveyor!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if github_actions:
                    try:
                        rotate_token_in_github_actions(
                            user, project, anaconda_token, token_name, gh
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on github actions!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

            except Exception as e:
                if "DEBUG_ANACONDA_TOKENS" in os.environ:
                    raise e
                failed = True

    if failed:
        if err_msg:
            raise RuntimeError(err_msg)
        else:
            raise RuntimeError(
                f"Rotating the feedstock token in providers for {user}/{project} failed!"
                " Try the command locally with DEBUG_ANACONDA_TOKENS"
                " defined in the environment to investigate!"
            )


def rotate_token_in_circle(user, project, binstar_token, token_name):
    from conda_smithy.ci_register import circle_token

    url_template = (
        "https://circleci.com/api/v1.1/project/github/{user}/{project}/envvar{extra}?"
        "circle-token={token}"
    )

    r = requests.get(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        )
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    for evar in r.json():
        if evar["name"] == token_name:
            have_binstar_token = True

    if have_binstar_token:
        r = requests.delete(
            url_template.format(
                token=circle_token,
                user=user,
                project=project,
                extra=f"/{token_name}",
            )
        )
        if r.status_code != 200:
            r.raise_for_status()

    data = {"name": token_name, "value": binstar_token}
    response = requests.post(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        ),
        data,
    )
    if response.status_code != 201:
        raise ValueError(response)


def rotate_token_in_drone(
    user, project, binstar_token, token_name, drone_endpoint
):
    from conda_smithy.ci_register import drone_session

    session = drone_session(drone_endpoint)

    r = session.get(f"/api/repos/{user}/{project}/secrets")
    r.raise_for_status()
    have_binstar_token = False
    for secret in r.json():
        if token_name == secret["name"]:
            have_binstar_token = True

    if have_binstar_token:
        r = session.patch(
            f"/api/repos/{user}/{project}/secrets/{token_name}",
            json={"data": binstar_token, "pull_request": False},
        )
        r.raise_for_status()
    else:
        response = session.post(
            f"/api/repos/{user}/{project}/secrets",
            json={
                "name": token_name,
                "data": binstar_token,
                "pull_request": False,
            },
        )
        if response.status_code != 200:
            response.raise_for_status()


def rotate_token_in_travis(
    user, project, feedstock_config_path, binstar_token, token_name
):
    """update the binstar token in travis."""
    from conda_smithy.ci_register import (
        travis_endpoint,
        travis_get_repo_info,
        travis_headers,
    )

    headers = travis_headers()

    repo_info = travis_get_repo_info(user, project)
    repo_id = repo_info["id"]

    r = requests.get(
        f"{travis_endpoint}/repo/{repo_id}/env_vars",
        headers=headers,
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    ev_id = None
    for ev in r.json()["env_vars"]:
        if ev["name"] == token_name:
            have_binstar_token = True
            ev_id = ev["id"]

    data = {
        "env_var.name": token_name,
        "env_var.value": binstar_token,
        "env_var.public": "false",
    }

    if have_binstar_token:
        r = requests.patch(
            f"{travis_endpoint}/repo/{repo_id}/env_var/{ev_id}",
            headers=headers,
            json=data,
        )
        r.raise_for_status()
    else:
        r = requests.post(
            f"{travis_endpoint}/repo/{repo_id}/env_vars",
            headers=headers,
            json=data,
        )
        if r.status_code != 201:
            r.raise_for_status()

    # we remove the token in the conda-forge.yml since on travis the
    # encrypted values override any value we put in the API
    with update_conda_forge_config(feedstock_config_path) as code:
        if (
            "travis" in code
            and "secure" in code["travis"]
            and token_name in code["travis"]["secure"]
        ):
            del code["travis"]["secure"][token_name]

            if len(code["travis"]["secure"]) == 0:
                del code["travis"]["secure"]

            if len(code["travis"]) == 0:
                del code["travis"]

            print(
                "An old value of the variable %s for travis was found in the "
                "conda-forge.yml. You may need to rerender this feedstock to "
                "use the new value since encrypted secrets inserted in travis.yml "
                "files override those set in the UI/API!"
            )


def rotate_token_in_azure(user, project, binstar_token, token_name):
    from vsts.build.v4_1.models import BuildDefinitionVariable

    from conda_smithy.azure_ci_utils import (
        build_client,
        get_default_build_definition,
    )
    from conda_smithy.azure_ci_utils import default_config as config

    bclient = build_client()

    existing_definitions = bclient.get_definitions(
        project=config.project_name, name=project
    )
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
    else:
        raise RuntimeError(
            f"Cannot add {token_name} to a repo that is not already registerd on azure CI!"
        )

    ed = bclient.get_definition(ed.id, project=config.project_name)

    if not hasattr(ed, "variables") or ed.variables is None:
        variables = {}
    else:
        variables = ed.variables

    variables[token_name] = BuildDefinitionVariable(
        allow_override=False,
        is_secret=True,
        value=binstar_token,
    )

    build_definition = get_default_build_definition(
        user,
        project,
        config=config,
        variables=variables,
        id=ed.id,
        revision=ed.revision,
    )

    bclient.update_definition(
        definition=build_definition,
        definition_id=ed.id,
        project=ed.project.name,
    )


def rotate_token_in_appveyor(feedstock_config_path, binstar_token, token_name):
    from conda_smithy.ci_register import appveyor_token

    headers = {"Authorization": f"Bearer {appveyor_token}"}
    url = "https://ci.appveyor.com/api/account/encrypt"
    response = requests.post(
        url, headers=headers, data={"plainValue": binstar_token}
    )
    if response.status_code != 200:
        raise ValueError(response)

    with update_conda_forge_config(feedstock_config_path) as code:
        code.setdefault("appveyor", {}).setdefault("secure", {})[
            token_name
        ] = response.content.decode("utf-8")


def rotate_token_in_github_actions(
    user, project, binstar_token, token_name, gh
):
    repo = gh.get_repo(f"{user}/{project}")
    assert repo.create_secret(token_name, binstar_token)
