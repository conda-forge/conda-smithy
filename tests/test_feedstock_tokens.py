import os
import json
from unittest import mock

import pytest
import scrypt

from conda_smithy.feedstock_tokens import (
    generate_and_write_feedstock_token,
    read_feedstock_token,
    feedstock_token_exists,
    register_feedstock_token,
    register_feedstock_token_with_proviers,
    is_valid_feedstock_token,
)


@pytest.mark.parametrize("project", ["bar", "bar-feedstock"])
@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_feedstock_tokens_roundtrip(
    gh_mock, git_mock, tmp_mock, tmpdir, repo, project
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )

    user = "foo"
    pth = os.path.expanduser("~/.conda-smithy/foo_bar_feedstock.token")
    token_json_pth = os.path.join(tmpdir, "tokens", "bar.json")
    os.makedirs(os.path.join(tmpdir, "tokens"), exist_ok=True)

    try:
        generate_and_write_feedstock_token(user, project)
        assert os.path.exists(pth)

        register_feedstock_token(user, project, repo)
        assert os.path.exists(token_json_pth)

        with open(pth, "r") as fp:
            feedstock_token = fp.read().strip()

        retval = is_valid_feedstock_token(user, project, feedstock_token, repo)
    finally:
        if os.path.exists(pth):
            os.remove(pth)

    assert retval


@pytest.mark.parametrize("project", ["bar", "bar-feedstock"])
@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_is_valid_feedstock_token_nofile(
    gh_mock, git_mock, tmp_mock, tmpdir, repo, project
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )

    user = "conda-forge"
    feedstock_token = "akdjhfl"
    retval = is_valid_feedstock_token(user, project, feedstock_token, repo)
    assert not retval


@pytest.mark.parametrize("project", ["bar", "bar-feedstock"])
@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_is_valid_feedstock_token_badtoken(
    gh_mock, git_mock, tmp_mock, tmpdir, repo, project
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )

    user = "conda-forge"
    feedstock_token = "akdjhfl"

    token_pth = os.path.join(tmpdir, "tokens", "bar.json")
    os.makedirs(os.path.dirname(token_pth), exist_ok=True)
    with open(token_pth, "w") as fp:
        json.dump({"salt": b"adf".hex(), "hashed_token": b"fgh".hex()}, fp)

    retval = is_valid_feedstock_token(user, project, feedstock_token, repo)
    assert not retval


def test_generate_and_write_feedstock_token():
    user = "bar"
    repo = "foo"

    pth = os.path.expanduser("~/.conda-smithy/bar_foo_feedstock.token")

    try:
        generate_and_write_feedstock_token(user, repo)

        assert os.path.exists(pth)

        # we cannot do it twice
        with pytest.raises(RuntimeError):
            generate_and_write_feedstock_token(user, repo)
    finally:
        if os.path.exists(pth):
            os.remove(pth)


def test_read_feedstock_token():
    user = "bar"
    repo = "foo"
    pth = os.path.expanduser("~/.conda-smithy/bar_foo_feedstock.token")

    # no token
    token, err = read_feedstock_token(user, repo)
    assert "No token found in" in err
    assert token is None

    # empty
    try:
        os.system("touch " + pth)
        token, err = read_feedstock_token(user, repo)
        assert "Empty token found in" in err
        assert token is None
    finally:
        if os.path.exists(pth):
            os.remove(pth)

    # token ok
    try:
        generate_and_write_feedstock_token(user, repo)
        token, err = read_feedstock_token(user, repo)
        assert err is None
        assert token is not None

    finally:
        if os.path.exists(pth):
            os.remove(pth)


@pytest.mark.parametrize("retval", [True, False])
@pytest.mark.parametrize("project", ["bar", "bar-feedstock"])
@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_feedstock_token_exists(
    gh_mock, git_mock, tmp_mock, tmpdir, repo, project, retval
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )

    user = "foo"
    os.makedirs(os.path.join(tmpdir, "tokens"), exist_ok=True)
    if retval:
        with open(os.path.join(tmpdir, "tokens", "bar.json"), "w") as fp:
            fp.write("blarg")

    assert feedstock_token_exists(user, project, repo) is retval

    assert git_mock.Repo.clone_from.called_with(
        "abc123", str(tmpdir), depth=1,
    )


@pytest.mark.parametrize("project", ["bar", "bar-feedstock"])
@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_feedstock_token_raises(
    gh_mock, git_mock, tmp_mock, tmpdir, repo, project
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )

    git_mock.Repo.clone_from.side_effect = ValueError("blarg")
    user = "foo"
    os.makedirs(os.path.join(tmpdir, "tokens"), exist_ok=True)
    with open(os.path.join(tmpdir, "tokens", "bar.json"), "w") as fp:
        fp.write("blarg")

    with pytest.raises(RuntimeError) as e:
        feedstock_token_exists(user, project, repo)

    assert "Testing for the feedstock token for" in str(e.value)

    assert git_mock.Repo.clone_from.called_with(
        "abc123", str(tmpdir), depth=1,
    )


@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.secrets")
@mock.patch("conda_smithy.feedstock_tokens.os.urandom")
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_register_feedstock_token_works(
    gh_mock, git_mock, tmp_mock, osuran_mock, secrets_mock, tmpdir, repo
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )
    secrets_mock.token_hex.return_value = "fgh"
    osuran_mock.return_value = b"\x80SA"

    user = "foo"
    project = "bar"
    os.makedirs(os.path.join(tmpdir, "tokens"), exist_ok=True)
    pth = os.path.expanduser("~/.conda-smithy/foo_bar_feedstock.token")
    token_json_pth = os.path.join(tmpdir, "tokens", "bar.json")

    try:
        generate_and_write_feedstock_token(user, project)

        register_feedstock_token(user, project, repo)

    finally:
        if os.path.exists(pth):
            os.remove(pth)

    assert git_mock.Repo.clone_from.called_with(
        "abc123", str(tmpdir), depth=1,
    )

    repo = git_mock.Repo.clone_from.return_value
    assert repo.index.add.called_with(token_json_pth)
    assert repo.index.commit.called_with(
        "added token for %s/%s" % (user, project)
    )
    assert repo.remote.return_value.pull.called_with(rebase=True)
    assert repo.remote.return_value.push.called_with()

    salted_token = scrypt.hash("fgh", b"\x80SA", buflen=256)
    data = {
        "salt": b"\x80SA".hex(),
        "hashed_token": salted_token.hex(),
    }

    with open(token_json_pth, "r") as fp:
        assert json.load(fp) == data


@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.secrets")
@mock.patch("conda_smithy.feedstock_tokens.os.urandom")
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_register_feedstock_token_notoken(
    gh_mock, git_mock, tmp_mock, osuran_mock, secrets_mock, tmpdir, repo
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )
    secrets_mock.token_hex.return_value = "fgh"
    osuran_mock.return_value = b"\x80SA"

    user = "foo"
    project = "bar"
    os.makedirs(os.path.join(tmpdir, "tokens"), exist_ok=True)
    pth = os.path.expanduser("~/.conda-smithy/foo_bar_feedstock.token")
    token_json_pth = os.path.join(tmpdir, "tokens", "bar.json")

    try:
        with pytest.raises(RuntimeError) as e:
            register_feedstock_token(user, project, repo)
    finally:
        if os.path.exists(pth):
            os.remove(pth)

    git_mock.Repo.clone_from.assert_not_called()

    repo = git_mock.Repo.clone_from.return_value
    repo.index.add.assert_not_called()
    repo.index.commit.assert_not_called()
    repo.remote.return_value.pull.assert_not_called()
    repo.remote.return_value.push.assert_not_called()

    assert not os.path.exists(token_json_pth)

    assert "No token found in" in str(e.value)


@pytest.mark.parametrize(
    "repo", ["GITHUB_TOKEN", "${GITHUB_TOKEN}", "GH_TOKEN", "${GH_TOKEN}"]
)
@mock.patch("conda_smithy.feedstock_tokens.secrets")
@mock.patch("conda_smithy.feedstock_tokens.os.urandom")
@mock.patch("conda_smithy.feedstock_tokens.tempfile")
@mock.patch("conda_smithy.feedstock_tokens.git")
@mock.patch("conda_smithy.github.gh_token")
def test_register_feedstock_token_exists_already(
    gh_mock, git_mock, tmp_mock, osuran_mock, secrets_mock, tmpdir, repo
):
    gh_mock.return_value = "abc123"
    tmp_mock.TemporaryDirectory.return_value.__enter__.return_value = str(
        tmpdir
    )
    secrets_mock.token_hex.return_value = "fgh"
    osuran_mock.return_value = b"\x80SA"

    user = "foo"
    project = "bar"
    os.makedirs(os.path.join(tmpdir, "tokens"), exist_ok=True)
    pth = os.path.expanduser("~/.conda-smithy/foo_bar_feedstock.token")
    token_json_pth = os.path.join(tmpdir, "tokens", "bar.json")
    with open(token_json_pth, "w") as fp:
        fp.write("blarg")

    try:
        generate_and_write_feedstock_token(user, project)

        with pytest.raises(RuntimeError) as e:
            register_feedstock_token(user, project, repo)

    finally:
        if os.path.exists(pth):
            os.remove(pth)

    assert git_mock.Repo.clone_from.called_with(
        "abc123", str(tmpdir), depth=1,
    )

    repo = git_mock.Repo.clone_from.return_value
    repo.index.add.assert_not_called()
    repo.index.commit.assert_not_called()
    repo.remote.return_value.pull.assert_not_called()
    repo.remote.return_value.push.assert_not_called()

    assert "Token for repo foo/bar already exists!" in str(e.value)


@pytest.mark.parametrize("drone", [True, False])
@pytest.mark.parametrize("circle", [True, False])
@pytest.mark.parametrize("azure", [True, False])
@pytest.mark.parametrize("travis", [True, False])
@pytest.mark.parametrize("clobber", [True, False])
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_drone")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_circle")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_travis")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_azure")
def test_register_feedstock_token_with_proviers(
    azure_mock,
    travis_mock,
    circle_mock,
    drone_mock,
    drone,
    circle,
    travis,
    azure,
    clobber,
):
    user = "foo"
    project = "bar"

    pth = os.path.expanduser("~/.conda-smithy/foo_bar_feedstock.token")

    try:
        generate_and_write_feedstock_token(user, project)
        feedstock_token, _ = read_feedstock_token(user, project)

        register_feedstock_token_with_proviers(
            user,
            project,
            drone=drone,
            circle=circle,
            travis=travis,
            azure=azure,
            clobber=clobber,
        )

        if drone:
            assert drone_mock.called_with(
                user, project, feedstock_token, clobber
            )
        else:
            drone_mock.assert_not_called()

        if circle:
            assert circle_mock.called_with(
                user, project, feedstock_token, clobber
            )
        else:
            circle_mock.assert_not_called()

        if travis:
            assert travis_mock.called_with(
                user, project, feedstock_token, clobber
            )
        else:
            travis_mock.assert_not_called()

        if azure:
            assert azure_mock.called_with(
                user, project, feedstock_token, clobber
            )
        else:
            azure_mock.assert_not_called()
    finally:
        if os.path.exists(pth):
            os.remove(pth)


@pytest.mark.parametrize("drone", [True, False])
@pytest.mark.parametrize("circle", [True, False])
@pytest.mark.parametrize("azure", [True, False])
@pytest.mark.parametrize("travis", [True, False])
@pytest.mark.parametrize("clobber", [True, False])
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_drone")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_circle")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_travis")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_azure")
def test_register_feedstock_token_with_proviers_notoken(
    azure_mock,
    travis_mock,
    circle_mock,
    drone_mock,
    drone,
    circle,
    travis,
    azure,
    clobber,
):
    user = "foo"
    project = "bar"

    with pytest.raises(RuntimeError) as e:
        register_feedstock_token_with_proviers(
            user,
            project,
            drone=drone,
            circle=circle,
            travis=travis,
            azure=azure,
            clobber=clobber,
        )

    assert "No token" in str(e.value)

    drone_mock.assert_not_called()
    circle_mock.assert_not_called()
    travis_mock.assert_not_called()
    azure_mock.assert_not_called()


@pytest.mark.parametrize("provider", ["drone", "circle", "travis", "azure"])
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_drone")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_circle")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_travis")
@mock.patch("conda_smithy.feedstock_tokens.add_feedstock_token_to_azure")
def test_register_feedstock_token_with_proviers_error(
    azure_mock, travis_mock, circle_mock, drone_mock, provider,
):
    user = "foo"
    project = "bar-feedstock"

    pth = os.path.expanduser("~/.conda-smithy/foo_bar_feedstock.token")

    if provider == "drone":
        drone_mock.side_effect = ValueError("blah")
    if provider == "circle":
        circle_mock.side_effect = ValueError("blah")
    if provider == "travis":
        travis_mock.side_effect = ValueError("blah")
    if provider == "azure":
        azure_mock.side_effect = ValueError("blah")

    try:
        generate_and_write_feedstock_token(user, project)
        feedstock_token, _ = read_feedstock_token(user, project)

        with pytest.raises(RuntimeError) as e:
            register_feedstock_token_with_proviers(
                user, project,
            )

        assert "on %s" % provider in str(e.value)
    finally:
        if os.path.exists(pth):
            os.remove(pth)
