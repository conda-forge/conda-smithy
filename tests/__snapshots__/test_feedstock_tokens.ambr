# serializer version: 1
# name: test_feedstock_token_raises[$GH_TOKEN-bar-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GH_TOKEN-bar-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GH_TOKEN-bar-feedstock-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GH_TOKEN-bar-feedstock-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GITHUB_TOKEN-bar-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GITHUB_TOKEN-bar-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GITHUB_TOKEN-bar-feedstock-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[$GITHUB_TOKEN-bar-feedstock-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GH_TOKEN}-bar-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GH_TOKEN}-bar-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GH_TOKEN}-bar-feedstock-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GH_TOKEN}-bar-feedstock-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GITHUB_TOKEN}-bar-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GITHUB_TOKEN}-bar-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GITHUB_TOKEN}-bar-feedstock-None]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
# name: test_feedstock_token_raises[${GITHUB_TOKEN}-bar-feedstock-azure]
  FeedstockTokenError('Testing for the feedstock token for foo/bar-feedstock on provider azure failed! Try the command locally with DEBUG_FEEDSTOCK_TOKENS defined in the environment to investigate!')
# ---
